Index dataframes tables

mmkekic opened the pull request:
The PR adds option to flag tables columns to be indexed.

mmkekic committed:
Add tests for columns_to_index kwarg in store_pandas_to_tables

The tests check: all elements in function argument are in
tables.attrs.columns_to_index; KeyError with column name is raised if input
list is not subset of dataframe columns; the argument is
not given nothing is flagged.

mmkekic committed:
Add docstrings for store_pandas_as_tables

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -89,3 +104,9 @@ def store_pandas_as_tables(h5out              : tb.file.File ,
         table.append(arr)
         table.flush()
+
+    if columns_to_index:
```
```suggestion
    if columns_to_index is not None:
```

gonzaponte commented on invisible_cities/io/dst_io_test.py:
```diff
@@ -188,3 +191,35 @@ def test_store_pandas_as_tables_unordered_df(config_tmpdir, df1, df2):
     df_read = load_dst(filename, group_name, table_name)
     assert_dataframes_equal(df_read, pd.concat([df1, df2], sort=True).reset_index(drop=True))
+
+
```
Another good test could be opening the file and checking that the table is actually indexed:

`a_table.indexed` is a property that indicates whether a table is indexed
`a_table.cols.<col_name>.indexed` is a property that indicates whether a specific column is indexed.

mmkekic replied:
But this is just labeling columns to be indexed, is not indexing them (thats whats done at the end of the city for example using  `index_tables`function of components)
We could index table  in store_pandas_as_tables but in case it is called repeatedly inside cities it doesn't seem efficient.

gonzaponte replied:
We have io/indexation_test.py that creates a dummy city and ensures every writer indexes the tables it writes. This should be added there. However this writer has a slightly different structure than the other, so it may require some adaptation.

We also need another PR to index tables in the cities that use this writer.

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -89,3 +104,9 @@ def store_pandas_as_tables(h5out              : tb.file.File ,
+        if set(columns_to_index).issubset(set(df.columns)):
+            table.set_attr('columns_to_index', columns_to_index)
+        else:
+            raise KeyError('column to index not present in the dataframe')
```
Clarifying which column failed would help the user.

gonzaponte reviewed (changes requested):
Cool. I would like just one more test to be added.

mmkekic committed:
Add test to check load_dst converts from bytes

The test where dataframe is read from
Kr83_full_nexus_v5_03_01_ACTIVE_7bar_1evt.sim.h5 file that contains
strings is added, and the test_store_pandas_as_tables_exact does not
decode explicitly.

mmkekic committed:
Change load_dst to decode byte type to strings

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -9,14 +9,20 @@
 from    typing             import Optional
+from    typing             import Sequence
+
+def _decode_df_to_str(df):
```
how about `decode_str_columns`?

gonzaponte commented on invisible_cities/io/dst_io_test.py:
```diff
@@ -98,6 +101,12 @@ def test_load_dsts_warns_if_not_existing_file(ICDATADIR):
     with pytest.warns(UserWarning, match='does not exist'):
         load_dsts([good_file, wrong_file], group, node)
 
+def test_load_dst_converts_from_bytest(ICDATADIR, fixed_dataframe):
```
```suggestion
def test_load_dst_converts_from_bytes(ICDATADIR, fixed_dataframe):
```
small typo

mmkekic committed:
Add store_pandas_to_tables writer in indexation_test

Since the writer has different structure than others the auxiliary
_df_writer had to be defined

mmkekic committed:
Index columns in all esmeralda writers

mmkekic committed:
Index event column in beersheba deconv_writer

mmkekic committed:
Add columns_to_index in store_pandas_as_tables

The writer is now labeling columns_to_be_index. The table is not
indexed since it should be done after the file is fully written.

mmkekic committed:
Rename store_pandas_as_tables to df_writer

mmkekic committed:
Cosmetis of alingment in load_dst_io_test

gonzaponte reviewed (approved):
An improved version of the dataframe writer with table indexation. The writer is now renamed to `df_writer`, which is much more comfortable. This PR also includes the indexation of the `event` column in the output of Beersheva and Esmeralda.

This change suggests that we should ensure each city indexes its output properly, but that is left for a different PR. 

The new feature is properly and thoughtfully tested. Nice job!
